{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4d283f45-31d6-406e-a794-7fe34bd1cec5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# install neurokit2 (delete after successfull installation)\n",
    "#pip install neurokit2\n",
    "\n",
    "import neurokit2 as nk\n",
    "from pathlib import Path\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from numpy import nan\n",
    "from itertools import chain\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "85091f7b-cdb1-45a1-9f7c-596febd2e211",
   "metadata": {},
   "outputs": [
    {
     "ename": "IndexError",
     "evalue": "list index out of range",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mIndexError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[1;32m~\\AppData\\Local\\Temp\\ipykernel_19900\\823684411.py\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m      8\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      9\u001b[0m \u001b[1;31m# load ecg data ('.txt') as df\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 10\u001b[1;33m \u001b[0mecg\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mfpaths\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;36m0\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     11\u001b[0m \u001b[0mecg\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mecg\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mread_text\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m','\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m-\u001b[0m\u001b[1;36m1\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     12\u001b[0m \u001b[0mecg\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mnp\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0marray\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mfloat\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mi\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;32mfor\u001b[0m \u001b[0mi\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mecg\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mreshape\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mlen\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mecg\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;36m1\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mIndexError\u001b[0m: list index out of range"
     ]
    }
   ],
   "source": [
    "# get file paths of '.txt' files in CURRENT folder\n",
    "fpaths=[i for i in Path.cwd().glob('*.txt')]\n",
    "\n",
    "# care about icg later, do ecg first!\n",
    "# icg=fpaths[0]\n",
    "# icg = icg.read_text()\n",
    "# icg=[float(i) for i in icg.split(',')[:-1]]\n",
    "\n",
    "# load ecg data ('.txt') as df\n",
    "ecg=fpaths[1] #Note Stefan: Bei mir kam hier ein IndexError, weil [1] nicht existiert, aber auch [0] existiert nicht. Wurde der Datensatz nicht eingelesen?\n",
    "ecg = ecg.read_text().split(',')[:-1]\n",
    "ecg=np.array([float(i) for i in ecg]).reshape((len(ecg), 1))\n",
    "ecg=pd.DataFrame(ecg, columns=['ECG'])\n",
    "\n",
    "# run r-peak detection on ecg data:\n",
    "peaks, info = nk.ecg_peaks(ecg, sampling_rate=1000)\n",
    "peak_times=np.where(peaks['ECG_R_Peaks']==1)[0]\n",
    "\n",
    "# plot ecg data with r peaks:\n",
    "plt.figure(figsize=(40,10))\n",
    "for i in peak_times:\n",
    "    plt.axvline(i, color='red')\n",
    "plt.plot(ecg)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "36a1aa73-358c-456a-a2e5-63f93eab787c",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'ecg' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m~\\AppData\\Local\\Temp\\ipykernel_19900\\1225060371.py\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m      1\u001b[0m \u001b[1;31m# detect artifacts\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      2\u001b[0m \u001b[0martifact_times\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 3\u001b[1;33m \u001b[1;32mfor\u001b[0m \u001b[0mi\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mrange\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;36m1\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mlen\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mecg\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m-\u001b[0m\u001b[1;36m1\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m      4\u001b[0m     \u001b[1;32mif\u001b[0m \u001b[0mecg\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0miloc\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mi\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;36m0\u001b[0m\u001b[1;33m]\u001b[0m \u001b[1;33m>\u001b[0m \u001b[0mecg\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0miloc\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mi\u001b[0m\u001b[1;33m-\u001b[0m\u001b[1;36m1\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;36m0\u001b[0m\u001b[1;33m]\u001b[0m \u001b[1;32mand\u001b[0m \u001b[0mecg\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0miloc\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mi\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;36m0\u001b[0m\u001b[1;33m]\u001b[0m \u001b[1;33m>\u001b[0m \u001b[0mecg\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0miloc\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mi\u001b[0m\u001b[1;33m+\u001b[0m\u001b[1;36m1\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;36m0\u001b[0m\u001b[1;33m]\u001b[0m \u001b[1;32mand\u001b[0m \u001b[0mecg\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0miloc\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mi\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;36m0\u001b[0m\u001b[1;33m]\u001b[0m \u001b[1;33m>\u001b[0m \u001b[1;36m0.5\u001b[0m\u001b[1;33m*\u001b[0m\u001b[0mmax\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mecg\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0miloc\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;36m0\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      5\u001b[0m         \u001b[0martifact_times\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mi\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mNameError\u001b[0m: name 'ecg' is not defined"
     ]
    }
   ],
   "source": [
    "# detect artifacts\n",
    "artifact_times=[]\n",
    "for i in range(1,len(ecg)-1):\n",
    "    if ecg.iloc[i, 0] > ecg.iloc[i-1, 0] and ecg.iloc[i, 0] > ecg.iloc[i+1, 0] and ecg.iloc[i, 0] > 0.5*max(ecg.iloc[:,0]):\n",
    "        artifact_times.append(i)\n",
    "\n",
    "# plot artifacts\n",
    "for i in artifact_times:\n",
    "    plt.axvline(i, color='green')\n",
    "\n",
    "# mark artifacts\n",
    "ecg_marked=ecg.copy()\n",
    "ecg_marked.iloc[artifact_times, 0]=nan\n",
    "\n",
    "# create final df\n",
    "final_df=ecg_marked.copy()\n",
    "final_df['ECG_R_Peaks']=peaks['ECG_R_Peaks']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "d2c7ea61",
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "module 'neurokit2' has no attribute 'ecg_load'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[1;32m~\\AppData\\Local\\Temp\\ipykernel_19900\\3139736061.py\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m      7\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      8\u001b[0m \u001b[1;31m# Load EKG data\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 9\u001b[1;33m \u001b[0mecg\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0minfo\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mnk\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mecg_load\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'C:\\\\Users\\\\stefa\\\\OneDrive\\\\Dokumente\\\\Studie 2\\\\Python\\\\021_BED1_LA_LEER.RAW_SIGNALS.rawECG1'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0msampling_rate\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;36m1000\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     10\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     11\u001b[0m \u001b[1;31m# Preprocess the EKG data\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mAttributeError\u001b[0m: module 'neurokit2' has no attribute 'ecg_load'"
     ]
    }
   ],
   "source": [
    "import neurokit2 as nk\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "def plot_impedance_cardiography(data):\n",
    "    \"\"\"\n",
    "    Plots Impedance Cardiography data\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    data : array or list\n",
    "        The array or list containing the impedance cardiography data\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    # Preprocess the impedance cardiography data\n",
    "    icg_processed = nk.ecg_process(data, sampling_rate=500)\n",
    "\n",
    "    # Plot the impedance cardiography data\n",
    "    plt.figure(figsize=(40,10))\n",
    "    plt.plot(icg_processed[\"ICG\"])\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "28edf6ef",
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid syntax (1215924184.py, line 2)",
     "output_type": "error",
     "traceback": [
      "\u001b[1;36m  File \u001b[1;32m\"C:\\Users\\stefa\\AppData\\Local\\Temp\\ipykernel_19900\\1215924184.py\"\u001b[1;36m, line \u001b[1;32m2\u001b[0m\n\u001b[1;33m    Schreibe ein Pythonskript mit NeuroKit2, um EKG Rohdaten zu verarbeiten\u001b[0m\n\u001b[1;37m             ^\u001b[0m\n\u001b[1;31mSyntaxError\u001b[0m\u001b[1;31m:\u001b[0m invalid syntax\n"
     ]
    }
   ],
   "source": [
    "# ab hier beginnt der Chat GPT Code\n",
    "Schreibe ein Pythonskript mit NeuroKit2, um EKG Rohdaten zu verarbeiten\n",
    "\n",
    "import neurokit2 as nk\n",
    "\n",
    "# Load EKG data\n",
    "data, info = nk.data_load('path/to/ekg_data.csv', sampling_rate=1000)\n",
    "\n",
    "# Preprocess the EKG data\n",
    "ekg_processed = nk.ecg_process(data, sampling_rate=1000)\n",
    "\n",
    "# Extract Heart Rate Variability (HRV) features\n",
    "hrv_features = nk.ecg_hrv(ekg_processed[\"ECG\"], sampling_rate=1000)\n",
    "\n",
    "# Extract Heart Rate (HR) features\n",
    "hr_features = nk.ecg_hr(ekg_processed[\"ECG\"], sampling_rate=1000)\n",
    "\n",
    "# Extract QRS features\n",
    "qrs_features = nk.ecg_qrs(ekg_processed[\"ECG\"], sampling_rate=1000)\n",
    "\n",
    "Nutze den folgenden Code als Grundlage und schreibe aufbauend darauf ein Pythonskript, das Artefakte markiert\n",
    "pip install neurokit2\n",
    "\n",
    "import neurokit2 as nk\n",
    "from pathlib import Path\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from numpy import nan\n",
    "from itertools import chain\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "fpaths=[i for i in Path.cwd().glob('*.txt')]\n",
    "\n",
    "# care about icg later, do ecg first!\n",
    "# icg=fpaths[0]\n",
    "# icg = icg.read_text()\n",
    "# icg=[float(i) for i in icg.split(',')[:-1]]\n",
    "\n",
    "# load ecg data ('.txt') as df\n",
    "ecg=fpaths[1]\n",
    "ecg = ecg.read_text().split(',')[:-1]\n",
    "ecg=np.array([float(i) for i in ecg]).reshape((len(ecg), 1))\n",
    "ecg=pd.DataFrame(ecg, columns=['ECG'])\n",
    "\n",
    "# run r-peak detection on ecg data:\n",
    "peaks, info = nk.ecg_peaks(ecg, sampling_rate=1000)\n",
    "peak_times=np.where(peaks['ECG_R_Peaks']==1)[0]\n",
    "\n",
    "# plot ecg data with r peaks:\n",
    "plt.figure(figsize=(40,10))\n",
    "for i in peak_times:\n",
    "    plt.axvline(i, color='red')\n",
    "plt.plot(ecg)\n",
    "\n",
    "# detect artifacts\n",
    "artifact_times=[]\n",
    "for i in range(1,len(ecg)-1):\n",
    "    if ecg.iloc[i, 0] > ecg.iloc[i-1, 0] and ecg.iloc[i, 0] > ecg.iloc[i+1, 0] and ecg.iloc[i, 0] > 0.5*max(ecg.iloc[:,0]):\n",
    "        artifact_times.append(i)\n",
    "\n",
    "# plot artifacts\n",
    "for i in artifact_times:\n",
    "    plt.axvline(i, color='green')\n",
    "\n",
    "# mark artifacts\n",
    "ecg_marked=ecg.copy()\n",
    "ecg_marked.iloc[artifact_times, 0]=nan\n",
    "\n",
    "# create final df\n",
    "final_df=ecg_marked.copy()\n",
    "final_df['ECG_R_Peaks']=peaks['ECG_R_Peaks']\n",
    "\n",
    "# save final df\n",
    "final_df.to_csv('ecg_with_artifacts.csv')\n",
    "\n",
    "wie korrigiere ich diese Fehlermeldung?\n",
    "  File \"C:\\Users\\stefa\\AppData\\Local\\Temp\\ipykernel_19900\\2777292691.py\", line 2\n",
    "    pip install neurokit2\n",
    "        ^\n",
    "SyntaxError: invalid syntax\n",
    "\n",
    "\n",
    "Der Fehler liegt darin, dass du versuchst, ein Python-Skript auszuführen, indem du \"pip install neurokit2\" eingibst. Dies ist kein gültiger Python-Befehl. Wenn du NeuroKit2 installieren möchtest, musst du die Kommandozeile öffnen und den Befehl \"pip install neurokit2\" dort ausführen.\n",
    "Wie öffne ich die Kommandozeile?\n",
    "\n",
    "Um die Kommandozeile unter Windows zu öffnen, kannst du entweder die Tastenkombination Windows + R drücken und dann \"cmd\" eingeben, oder du kannst das Startmenü öffnen und dort \"cmd\" eingeben. Auf Mac kannst du die Tastenkombination Command + Space drücken und dann \"Terminal\" eingeben.\n",
    "\n",
    "Wie korrigiere ich diese Fehlermeldung?\n",
    "---------------------------------------------------------------------------\n",
    "IndexError                                Traceback (most recent call last)\n",
    "~\\AppData\\Local\\Temp\\ipykernel_19900\\2608316579.py in <module>\n",
    "      8 \n",
    "      9 # load ecg data ('.txt') as df\n",
    "---> 10 ecg=fpaths[1]\n",
    "     11 ecg = ecg.read_text().split(',')[:-1]\n",
    "     12 ecg=np.array([float(i) for i in ecg]).reshape((len(ecg), 1))\n",
    "\n",
    "IndexError: list index out of range\n",
    "\n",
    "\n",
    "Der Fehler liegt darin, dass du versuchst, auf ein Element des Arrays fpaths zuzugreifen, das nicht existiert. Du versuchst, auf das zweite Element zuzugreifen (fpaths[1]), aber das Array hat nur ein Element (fpaths[0]). Verändere die Indexzahl (1) so, dass sie dem Element entspricht, auf das du zugreifen möchtest.\n",
    "\n",
    "Wie sage ich Python, dass ich auf eine Datei aus diesem Ordner zugreifen möchte?\n",
    "C:\\Users\\stefa\\OneDrive\\Dokumente\\Studie 2\\Python\n",
    "\n",
    "Du kannst die Funktion pathlib.Path.cwd() verwenden, um die aktuelle Arbeitsverzeichnis zu bekommen. Dann kannst du die glob()-Funktion verwenden, um alle Dateien im Verzeichnis zu erhalten. Zum Beispiel:\n",
    "\n",
    "# import the pathlib library\n",
    "from pathlib import Path\n",
    "\n",
    "# get the current working directory\n",
    "cwd = Path.cwd()\n",
    "\n",
    "# get all files in the directory\n",
    "fpaths = [i for i in cwd.glob('*.txt')]\n",
    "\n",
    "Den Index von 1 auf 0 zu korrigieren hat nicht funktioniert. Wie könnte ich den Fehler korrigieren?\n",
    "---------------------------------------------------------------------------\n",
    "IndexError                                Traceback (most recent call last)\n",
    "~\\AppData\\Local\\Temp\\ipykernel_19900\\823684411.py in <module>\n",
    "      8 \n",
    "      9 # load ecg data ('.txt') as df\n",
    "---> 10 ecg=fpaths[0]\n",
    "     11 ecg = ecg.read_text().split(',')[:-1]\n",
    "     12 ecg=np.array([float(i) for i in ecg]).reshape((len(ecg), 1))\n",
    "\n",
    "IndexError: list index out of range\n",
    "\n",
    "\n",
    "Der Fehler liegt darin, dass du versuchst, auf ein Element des Arrays fpaths zuzugreifen, das nicht existiert. Der Index 0 bezeichnet das erste Element in diesem Array, aber es gibt kein erstes Element. Stattdessen musst du den Index ändern, so dass er dem Element entspricht, auf das du zugreifen möchtest. Wenn du auf das erste Element zugreifen möchtest, musst du also den Index 0 verwenden.\n",
    "\n",
    "Schreibe ein Pythonskript, das Artifakte im EKG Signal korrigiert\n",
    "\n",
    "import neurokit2 as nk\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from numpy import nan\n",
    "from itertools import chain\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Load EKG data\n",
    "data, info = nk.data_load('path/to/ekg_data.csv', sampling_rate=1000)\n",
    "\n",
    "# Preprocess the EKG data\n",
    "ekg_processed = nk.ecg_process(data, sampling_rate=1000)\n",
    "\n",
    "# Detect artifacts \n",
    "artifact_times=[]\n",
    "for i in range(1,len(ekg_processed[\"ECG\"])-1):\n",
    "    if ekg_processed[\"ECG\"][i] > ekg_processed[\"ECG\"][i-1] and ekg_processed[\"ECG\"][i] > ekg_processed[\"ECG\"][i+1] and ekg_processed[\"ECG\"][i] > 0.5*max(ekg_processed[\"ECG\"]):\n",
    "        artifact_times.append(i)\n",
    "\n",
    "# Replace artifacts with linear\n",
    "Wie kann ich diesen Fehler korrigieren?\n",
    "  File \"C:\\Users\\stefa\\AppData\\Local\\Temp\\ipykernel_19900\\4191037142.py\", line 9\n",
    "    data, info = nk.data_load('C:\\Users\\stefa\\OneDrive\\Dokumente\\Studie 2\\Python\\021_BED1_LA_LEER.RAW_SIGNALS.rawECG1', sampling_rate=1000)\n",
    "                                                                                                                      ^\n",
    "SyntaxError: (unicode error) 'unicodeescape' codec can't decode bytes in position 2-3: truncated \\UXXXXXXXX escape\n",
    "\n",
    "\n",
    "Der Fehler liegt darin, dass du versuchst, einen Dateipfad mit einem Backslash (\\) anzugeben, aber Python interpretiert das als ein Unicode-Escape-Zeichen (\\Uxxxxxxx). Um dies zu korrigieren, musst du den Backslash doppelt eingeben, damit Python weiß, dass du kein Unicode-Escape-Zeichen verwenden möchtest. Zum Beispiel:\n",
    "\n",
    "data, info = nk.data_load('C:\\\\Users\\\\stefa\\\\OneDrive\\\\Dokumente\\\\Studie 2\\\\Python\\\\021_BED1_LA_LEER.RAW_SIGNALS.rawECG1', sampling_rate=1000)\n",
    "\n",
    "Wie kann ich diesen Fehler korrigieren?\n",
    "import neurokit2 as nk\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "import numpy as np\n",
    "\n",
    "from numpy import nan\n",
    "\n",
    "from itertools import chain\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "​\n",
    "\n",
    "# Load EKG data\n",
    "\n",
    "data, info = nk.data_load('C:\\\\Users\\\\stefa\\\\OneDrive\\\\Dokumente\\\\Studie 2\\\\Python\\\\021_BED1_LA_LEER.RAW_SIGNALS.rawECG1', sampling_rate=1000)\n",
    "\n",
    "​\n",
    "\n",
    "# Preprocess the EKG data\n",
    "\n",
    "ekg_processed = nk.ecg_process(data, sampling_rate=1000)\n",
    "\n",
    "​\n",
    "\n",
    "# Detect artifacts \n",
    "\n",
    "artifact_times=[]\n",
    "\n",
    "for i in range(1,len(ekg_processed[\"ECG\"])-1):\n",
    "\n",
    "    if ekg_processed[\"ECG\"][i] > ekg_processed[\"ECG\"][i-1] and ekg_processed[\"ECG\"][i] > ekg_processed[\"ECG\"][i+1] and ekg_processed[\"ECG\"][i] > 0.5*max(ekg_processed[\"ECG\"]):\n",
    "\n",
    "        artifact_times.append(i)\n",
    "\n",
    "​\n",
    "\n",
    "# Replace artifacts with linear\n",
    "\n",
    "---------------------------------------------------------------------------\n",
    "AttributeError                            Traceback (most recent call last)\n",
    "~\\AppData\\Local\\Temp\\ipykernel_19900\\135997045.py in <module>\n",
    "      7 \n",
    "      8 # Load EKG data\n",
    "----> 9 data, info = nk.data_load('C:\\\\Users\\\\stefa\\\\OneDrive\\\\Dokumente\\\\Studie 2\\\\Python\\\\021_BED1_LA_LEER.RAW_SIGNALS.rawECG1', sampling_rate=1000)\n",
    "     10 \n",
    "     11 # Preprocess the EKG data\n",
    "\n",
    "AttributeError: module 'neurokit2' has no attribute 'data_load'\n",
    "\n",
    "\n",
    "Der Fehler liegt darin, dass du versuchst, die Funktion data_load() aus dem Modul neurokit2 aufzurufen, aber diese Funktion existiert nicht in diesem Modul. Stattdessen musst du die Funktion nk.ecg_load() verwenden. Zum Beispiel:\n",
    "\n",
    "# Load EKG data\n",
    "data, info = nk.ecg_load('C:\\\\Users\\\\stefa\\\\OneDrive\\\\Dokumente\\\\Studie 2\\\\Python\\\\021_BED1_LA_LEER.RAW_SIGNALS.rawECG1', sampling_rate=1000)\n",
    "\n",
    "Schreibe eine Funktion in Python, mit der Impedanzkardiographie abgebildet wird\n",
    "\n",
    "import neurokit2 as nk\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "def plot_impedance_cardiography(data):\n",
    "    \"\"\"\n",
    "    Plots Impedance Cardiography data\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    data : array or list\n",
    "        The array or list containing the impedance cardiography data\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    # Preprocess the impedance cardiography data\n",
    "    icg_processed = nk.ecg_process(data, sampling_rate=500)\n",
    "\n",
    "    # Plot the impedance cardiography data\n",
    "    plt.figure(figsize=(40,10))\n",
    "    plt.plot(icg_processed[\"ICG\"])\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10810d45",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
